### PR TITLE
Ignore duplicate consecutive history entries from the same url

### PR DESCRIPTION
**Pull request by jgkamat**
I'm finding that on newer versions of qt, I'm getting many duplicate entries in my history list, which is causing my history file to balloon. On a computer that I started using about ~9 months ago, my history file is ~250mb. This is due to a combination of:

1. https://github.com/qutebrowser/qutebrowser/issues/3596
2. We save history blindly every time.

This PR is a naiive solution (to be more similar to chrome's history list) which only saves into history if the last URL wasn't the exact same. At the very least, this avoids double saves on every url (which I'm getting at the moment). The main difference is that chrome updates the access time to be the newer access, while this dosen't. I think there's more value in the older time personally, and I'd rather keep things simple, so I'm trying to avoid implementing that.

Unfortunately, anchor changes will still get through (which is the main cause of my problem), but fixing that's a more intrusive change.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4699)

### CODE REVIEW

File: qutebrowser/browser/history.py [hunk 1]
```diff
@@ -148,6 +148,8 @@ def __init__(self, progress, parent=None):
                                       'redirect': 'NOT NULL'},
                          parent=parent)
         self._progress = progress
+        # Store the last saved url to avoid duplicate immedate saves.
+        self.last_url = None
```
Comment (The-Compiler) [inline qutebrowser/browser/history.py]: Should be private:

```suggestion
        self._last_url = None
```

**Comment by mschilli87** [comment]
Won't this make approaches like https://github.com/qutebrowser/qutebrowser/pull/4403 to resolve https://github.com/qutebrowser/qutebrowser/issues/601 more or less impossible?

**Comment by The-Compiler** [comment]
@mschilli87 I don't think so - this doesn't ignore multiple visits to the same URL in general - only if the same URL is loaded twice immediately after the first load.

**Comment by mschilli87** [comment]
I must have misread the

> I think there's more value in the older time personally, and I'd rather keep things simple, so I'm trying to avoid implementing that.

part then. :wink:

As long as it we still keep track of when a page was visted (and not just the last time), all should be good.

---
Just pinging @vchimishuk to make sure the two PR remain compatible.

File: tests/unit/browser/test_history.py [hunk 2]
```diff
@@ -229,6 +237,18 @@ def test_exclude(self, web_history, config_stub):
         assert list(web_history)
         assert not list(web_history.completion)
 
+    def test_no_immedate_duplicates(self, web_history, caplog, mock_time, mocker):
```
Comment (rcorre) [inline tests/unit/browser/test_history.py]: Can't this test just check `list(web_history)` like the rest? I don't see a reason to introduce mocks here.

**Comment by jgkamat** [comment]
Marcel Schilling writes:

>> Won't this make approaches like #4403 to resolve #601 more or less impossible?
> As long as it we still keep track of when a page was visted (and not just the last time), all should be good.

Yup, I don't think this should make that impossible for sure. It will change
the results though, holding down refresh won't give an extremely high score
anymore. New urls will have a lower score than 'they should' as well, because
double-urls have been stored for a while. The only change is that if two
consecutive urls that would have been stored are exactly the same, the later
ones are dropped. The comment is just saying that chrome 'drops' the previous
ones, and not the later ones.

Of course, it's pretty trivial to disable all this (just an extra flag to the
if), if we need to.

**Comment by jgkamat** [comment]
Florian Bruhin writes:

> Should we also take care of getting rid of those in the existing history file?

I don't think it's worth it to clean these out ourselves, and I personally
don't like the idea of deleting absolutely anything without explicit
confirmation (changing storage behavior is less intrusive though). Maybe
someone could make a tiny script to cleanup duplicates (I might do this to
trim down my completion history as well, as it's still rather slow for me). If
you really want to I could probably think up something, but I'd rather keep
this as simple as possible.

File: tests/unit/browser/test_history.py [hunk 3]
```diff
@@ -229,6 +229,37 @@ def test_exclude(self, web_history, config_stub):
         assert list(web_history)
         assert not list(web_history.completion)
 
+    def test_no_immedate_duplicates(self, web_history, mock_time):
+        url = QUrl("http://example.com")
+        url2 = QUrl("http://example2.com")
+        web_history.add_from_tab(QUrl(url), QUrl(url), 'title')
+        hist = list(web_history)
+        assert list(web_history)
```
Comment (The-Compiler) [inline tests/unit/browser/test_history.py]: Why do `list(web_history)` twice here? Couldn't this just be

```suggestion
        assert hist
```

instead?

File: tests/unit/browser/test_history.py [hunk 4]
```diff
@@ -229,6 +229,37 @@ def test_exclude(self, web_history, config_stub):
...
+        web_history.add_from_tab(QUrl(url), QUrl(url), 'title')
+        assert list(web_history) == hist
+        web_history.add_from_tab(QUrl(url2), QUrl(url2), 'title')
+        assert len(list(web_history)) != hist
```
Comment (The-Compiler) [inline tests/unit/browser/test_history.py]: This assertion will always be true, you're comparing a length with a list...

File: tests/unit/browser/test_history.py [hunk 5]
```diff
@@ -229,6 +229,37 @@ def test_exclude(self, web_history, config_stub):
...
+
+    def test_delete_add_tab(self, web_history, mock_time):
+        url = QUrl("http://example.com")
+        web_history.add_from_tab(QUrl(url), QUrl(url), 'title')
+        hist = list(web_history)
+        assert list(web_history)
```
Comment (The-Compiler) [inline tests/unit/browser/test_history.py]: same as above

File: tests/unit/browser/test_history.py [hunk 6]
```diff
@@ -229,6 +229,37 @@ def test_exclude(self, web_history, config_stub):
...
+        web_history.delete_url(QUrl(url))
+        assert len(web_history) == 0
+        web_history.add_from_tab(QUrl(url), QUrl(url), 'title')
+        assert list(web_history) == hist
+
+    def test_clear_add_tab(self, web_history, mock_time):
+        url = QUrl("http://example.com")
+        web_history.add_from_tab(QUrl(url), QUrl(url), 'title')
+        hist = list(web_history)
```
Comment (The-Compiler) [inline tests/unit/browser/test_history.py]: same as above

**Comment by The-Compiler** [comment]
Looks good as-is - I don't think there's a reason to disable the behavior, and you got me convinced on not touching the existing history. Thanks!

**Pull request merged**